In [23]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split,cross_val_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score,r2_score

## Using glass dataset

In [2]:
glass_data = pd.read_csv('glass.csv')
glass_data

RI     Na    Mg    Al     Si     K    Ca    Ba   Fe  Type
0    1.52101  13.64  4.49  1.10  71.78  0.06  8.75  0.00  0.0     1
1    1.51761  13.89  3.60  1.36  72.73  0.48  7.83  0.00  0.0     1
2    1.51618  13.53  3.55  1.54  72.99  0.39  7.78  0.00  0.0     1
3    1.51766  13.21  3.69  1.29  72.61  0.57  8.22  0.00  0.0     1
4    1.51742  13.27  3.62  1.24  73.08  0.55  8.07  0.00  0.0     1
..       ...    ...   ...   ...    ...   ...   ...   ...  ...   ...
209  1.51623  14.14  0.00  2.88  72.61  0.08  9.18  1.06  0.0     7
210  1.51685  14.92  0.00  1.99  73.06  0.00  8.40  1.59  0.0     7
211  1.52065  14.36  0.00  2.02  73.42  0.00  8.44  1.64  0.0     7
212  1.51651  14.38  0.00  1.94  73.61  0.00  8.48  1.57  0.0     7
213  1.51711  14.23  0.00  2.08  73.36  0.00  8.62  1.67  0.0     7

[214 rows x 10 columns]

In [5]:
glass_data.shape

(214, 10)

In [6]:
glass_data.describe()

RI          Na          Mg          Al          Si           K  \
count  214.000000  214.000000  214.000000  214.000000  214.000000  214.000000   
mean     1.518365   13.407850    2.684533    1.444907   72.650935    0.497056   
std      0.003037    0.816604    1.442408    0.499270    0.774546    0.652192   
min      1.511150   10.730000    0.000000    0.290000   69.810000    0.000000   
25%      1.516522   12.907500    2.115000    1.190000   72.280000    0.122500   
50%      1.517680   13.300000    3.480000    1.360000   72.790000    0.555000   
75%      1.519157   13.825000    3.600000    1.630000   73.087500    0.610000   
max      1.533930   17.380000    4.490000    3.500000   75.410000    6.210000   

               Ca          Ba          Fe        Type  
count  214.000000  214.000000  214.000000  214.000000  
mean     8.956963    0.175047    0.057009    2.780374  
std      1.423153    0.497219    0.097439    2.103739  
min      5.430000    0.000000    0.000000    1.000000  
25%      8.240000    0.000000    0.000000    1.000000  
50%      8.600000    0.000000    0.000000    2.000000  
75%      9.172500    0.000000    0.100000    3.000000  
max     16.190000    3.150000    0.510000    7.000000

In [7]:
glass_data.isna().sum()

RI      0
Na      0
Mg      0
Al      0
Si      0
K       0
Ca      0
Ba      0
Fe      0
Type    0
dtype: int64

In [8]:
X = glass_data.drop(['Type'],axis=1)
y = glass_data['Type']

In [10]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.20,random_state= 15,stratify=y)


In [11]:
glass_knn_model = KNeighborsClassifier()
glass_knn_model.fit(X_train,y_train)

KNeighborsClassifier()

In [12]:
y_pred = glass_knn_model.predict(X_test)
y_pred

array([2, 1, 2, 7, 1, 7, 1, 2, 5, 7, 5, 1, 2, 7, 7, 1, 2, 2, 1, 1, 1, 7,
       1, 1, 1, 3, 1, 1, 2, 1, 2, 1, 1, 2, 2, 2, 2, 2, 1, 5, 1, 7, 2],
      dtype=int64)

In [13]:
print("Accuracy score   :",round(accuracy_score(y_test,y_pred),4))
print("R2 Score         :",round(r2_score(y_test,y_pred),4))

Accuracy score   : 0.6279
R2 Score         : 0.6477


In [15]:
import warnings
warnings.filterwarnings('ignore')
neighbours = list(range(1,50,2))
cv_scores = []

for i in neighbours:
    Glass_knn_model = KNeighborsClassifier(n_neighbors = i)
    cv_score = cross_val_score(estimator = glass_knn_model,X=X,y=y,cv=10)
    cv_scores.append(cv_score.mean())

In [16]:
cv_scores

[0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455,
 0.6454545454545455]

In [17]:
neighbours[cv_scores.index(max(cv_scores))]


1

## As per above line to get better result K should be K=1

In [18]:
glass_knn_model = KNeighborsClassifier(n_neighbors=1)
glass_knn_model.fit(X_train,y_train)
y_pred = glass_knn_model.predict(X_test)
print("Accuracy score   :", round(accuracy_score(y_test,y_pred),4))
print("R2 Score         :",round(r2_score(y_test,y_pred),4))

Accuracy score   : 0.6744
R2 Score         : 0.7907


##  If we take K=5 then Accuracy is 0.6279
## If we take K=1 then Accuracy is 0.6744

# Using zoo data

In [3]:
zoo_data = pd.read_csv('Zoo.csv')
zoo_data

animal name  hair  feathers  eggs  milk  airborne  aquatic  predator  \
0      aardvark     1         0     0     1         0        0         1   
1      antelope     1         0     0     1         0        0         0   
2          bass     0         0     1     0         0        1         1   
3          bear     1         0     0     1         0        0         1   
4          boar     1         0     0     1         0        0         1   
..          ...   ...       ...   ...   ...       ...      ...       ...   
96      wallaby     1         0     0     1         0        0         0   
97         wasp     1         0     1     0         1        0         0   
98         wolf     1         0     0     1         0        0         1   
99         worm     0         0     1     0         0        0         0   
100        wren     0         1     1     0         1        0         0   

     toothed  backbone  breathes  venomous  fins  legs  tail  domestic  \
0          1         1         1         0     0     4     0         0   
1          1         1         1         0     0     4     1         0   
2          1         1         0         0     1     0     1         0   
3          1         1         1         0     0     4     0         0   
4          1         1         1         0     0     4     1         0   
..       ...       ...       ...       ...   ...   ...   ...       ...   
96         1         1         1         0     0     2     1         0   
97         0         0         1         1     0     6     0         0   
98         1         1         1         0     0     4     1         0   
99         0         0         1         0     0     0     0         0   
100        0         1         1         0     0     2     1         0   

     catsize  type  
0          1     1  
1          1     1  
2          0     4  
3          1     1  
4          1     1  
..       ...   ...  
96         1     1  
97         0     6  
98         1     1  
99         0     7  
100        0     2  

[101 rows x 18 columns]

In [20]:
zoo_data.describe()

hair    feathers        eggs        milk    airborne     aquatic  \
count  101.000000  101.000000  101.000000  101.000000  101.000000  101.000000   
mean     0.425743    0.198020    0.584158    0.405941    0.237624    0.356436   
std      0.496921    0.400495    0.495325    0.493522    0.427750    0.481335   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    1.000000    0.000000    0.000000    0.000000   
75%      1.000000    0.000000    1.000000    1.000000    0.000000    1.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

         predator     toothed    backbone    breathes    venomous        fins  \
count  101.000000  101.000000  101.000000  101.000000  101.000000  101.000000   
mean     0.554455    0.603960    0.821782    0.792079    0.079208    0.168317   
std      0.499505    0.491512    0.384605    0.407844    0.271410    0.376013   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    1.000000    1.000000    0.000000    0.000000   
50%      1.000000    1.000000    1.000000    1.000000    0.000000    0.000000   
75%      1.000000    1.000000    1.000000    1.000000    0.000000    0.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

             legs        tail    domestic     catsize        type  
count  101.000000  101.000000  101.000000  101.000000  101.000000  
mean     2.841584    0.742574    0.128713    0.435644    2.831683  
std      2.033385    0.439397    0.336552    0.498314    2.102709  
min      0.000000    0.000000    0.000000    0.000000    1.000000  
25%      2.000000    0.000000    0.000000    0.000000    1.000000  
50%      4.000000    1.000000    0.000000    0.000000    2.000000  
75%      4.000000    1.000000    0.000000    1.000000    4.000000  
max      8.000000    1.000000    1.000000    1.000000    7.000000

In [21]:
zoo_data.isnull().sum()

animal name    0
hair           0
feathers       0
eggs           0
milk           0
airborne       0
aquatic        0
predator       0
toothed        0
backbone       0
breathes       0
venomous       0
fins           0
legs           0
tail           0
domestic       0
catsize        0
type           0
dtype: int64

In [26]:
zoo_Label = LabelEncoder()
zoo_data['animal name'] = zoo_Label.fit_transform(zoo_data['animal name'])

In [27]:
zoo_data

animal name  hair  feathers  eggs  milk  airborne  aquatic  predator  \
0              0     1         0     0     1         0        0         1   
1              1     1         0     0     1         0        0         0   
2              2     0         0     1     0         0        1         1   
3              3     1         0     0     1         0        0         1   
4              4     1         0     0     1         0        0         1   
..           ...   ...       ...   ...   ...       ...      ...       ...   
96            95     1         0     0     1         0        0         0   
97            96     1         0     1     0         1        0         0   
98            97     1         0     0     1         0        0         1   
99            98     0         0     1     0         0        0         0   
100           99     0         1     1     0         1        0         0   

     toothed  backbone  breathes  venomous  fins  legs  tail  domestic  \
0          1         1         1         0     0     4     0         0   
1          1         1         1         0     0     4     1         0   
2          1         1         0         0     1     0     1         0   
3          1         1         1         0     0     4     0         0   
4          1         1         1         0     0     4     1         0   
..       ...       ...       ...       ...   ...   ...   ...       ...   
96         1         1         1         0     0     2     1         0   
97         0         0         1         1     0     6     0         0   
98         1         1         1         0     0     4     1         0   
99         0         0         1         0     0     0     0         0   
100        0         1         1         0     0     2     1         0   

     catsize  type  
0          1     1  
1          1     1  
2          0     4  
3          1     1  
4          1     1  
..       ...   ...  
96         1     1  
97         0     6  
98         1     1  
99         0     7  
100        0     2  

[101 rows x 18 columns]

In [28]:
X = zoo_data.drop(['type'],axis=1)
y = zoo_data['type']

In [29]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.20,random_state= 15,stratify=y)

In [30]:
zoo_knn_model = KNeighborsClassifier()
zoo_knn_model.fit(X_train,y_train)

KNeighborsClassifier()

In [31]:
y_pred = zoo_knn_model.predict(X_test)
y_pred

array([2, 2, 1, 1, 1, 2, 1, 1, 1, 1, 1, 7, 1, 1, 1, 1, 1, 4, 2, 1, 1],
      dtype=int64)

In [32]:
print("Accuracy score   : ", round(accuracy_score(y_test,y_pred),4))
print("R2 Score         :",round(r2_score(y_test,y_pred),4))

Accuracy score   :  0.4286
R2 Score         : -0.8504


In [33]:
neighbours = list(range(1,50,2))
cv_scores = []

for i in neighbours:
    zoo_knn_model = KNeighborsClassifier(n_neighbors = i)
    cv_score = cross_val_score(estimator = zoo_knn_model,X=X,y=y,cv=10)
    cv_scores.append(cv_score.mean())

In [34]:
cv_scores

[0.4845454545454545,
 0.4445454545454545,
 0.4154545454545454,
 0.3854545454545455,
 0.3354545454545455,
 0.35545454545454547,
 0.3654545454545454,
 0.38545454545454544,
 0.37545454545454543,
 0.3854545454545455,
 0.3954545454545455,
 0.3954545454545454,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546,
 0.40545454545454546]

In [35]:
neighbours[cv_scores.index(max(cv_scores))]

1

In [36]:
zoo_knn_model = KNeighborsClassifier(n_neighbors=1)
zoo_knn_model.fit(X_train,y_train)
y_pred = zoo_knn_model.predict(X_test)
print("Accuracy score   :", round(accuracy_score(y_test,y_pred),4))
print("R2 Score         :",round(r2_score(y_test,y_pred),4))

Accuracy score   : 0.6667
R2 Score         : 0.333


## If we take K=5 then Accuracy is 0.4286
## If we take K=1 then Accuracy is 0.6667